src/test/java/com/google/edith/LoginServletTest.java

```diff
@@ -0,0 +1,103 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.edith;
+
+import com.google.edith.servlets.LoginServlet;
+
+import com.google.appengine.api.users.User;
+import com.google.appengine.api.users.UserService;
+import com.google.appengine.api.users.UserServiceFactory;
+import com.google.appengine.tools.development.testing.LocalUserServiceTestConfig;
+import com.google.appengine.tools.development.testing.LocalServiceTestHelper;
+import com.google.common.collect.ImmutableMap;
+import org.junit.Before;
+import org.junit.Test;
+import org.mockito.Mock;
+import org.mockito.MockitoAnnotations;
+import org.mockito.ArgumentCaptor;
+import java.util.HashMap;
+import java.util.Map;
+import java.io.IOException;
+import java.io.StringWriter;
+import java.io.PrintWriter;
+import javax.servlet.ServletException;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
+import static org.mockito.Mockito.when;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.atLeast;
+
+public final class LoginServletTest {
+  private Map<String, Object> map = ImmutableMap
+            .of("com.google.appengine.api.users.UserService.user_id_key", "12345");
+
+  private final LocalServiceTestHelper loggedInTestHelper =
+        new LocalServiceTestHelper(new LocalUserServiceTestConfig())
+            .setEnvAttributes(map)
+            .setEnvIsLoggedIn(true)
+            .setEnvAuthDomain("gmail")
+            .setEnvIsAdmin(true)
+            .setEnvEmail("user@gmail.com");
+      
+  private final LocalServiceTestHelper loggedOutTestHelper =
+        new LocalServiceTestHelper(new LocalUserServiceTestConfig())
+            .setEnvIsLoggedIn(false);
+  
+  private final UserService userService = UserServiceFactory.getUserService();
+  private LoginServlet loginServlet;
+
+  @Mock
+  HttpServletRequest request;
+
+  @Mock
+  HttpServletResponse response;
+
+  @Before
+  public void setUp() throws Exception {
+    MockitoAnnotations.initMocks(this);
+    loginServlet = new LoginServlet();
+  }
+
+  @Test
+  // Check if the servlet returns with user information if logged-in.
+  public void checks_ifUserLoggedIn_returnsUserInfo() throws IOException, ServletException {
+    loggedInTestHelper.setUp();
+    assertTrue(userService.isUserLoggedIn());
+
+    StringWriter stringWriter = new StringWriter();
+    PrintWriter writer = new PrintWriter(stringWriter);
+    when(response.getWriter()).thenReturn(writer);
+
+    loginServlet.doGet(request, response);
+    verify(response, atLeast(1)).getWriter();
```

Comment:
It looks like we are just verifying the writer is called, which is definitely useful. Should we add another test to check that the output has the correct userinfo output as well? Since this is logic (albeit trivial) in the servlet class that seems untested. 

E.g. If I accidentally removed line 66 in the LoginServlet class, tests would pass, but they probably should not.

Reply:
Yeah, most definitely.
I have added tests to assert that the userinfo must contain all of the fields.

Reply:
Thanks for adding tests!
I think its good to note that although we are testing the same public method, we are actually testing two different code paths and behaviors. One being that response.getWriter is called and another being UserInfo is populated correctly. I could imagine a world where one behavior succeeds and another fails, which indicates that we should probably separate these into two separate tests.

Although they might have some similar initial setup, we want to optimize for clean tests and isolating failure in so it's ok. Could we break out into two tests?

Reply:
Thanks for the suggestions. I have added another tests to just check the user info.
Changes made.